flang/lib/Optimizer/Analysis/AliasAnalysis.cpp

```diff
@@ -399,19 +404,30 @@ AliasAnalysis::Source AliasAnalysis::getSource(mlir::Value v) {
   if (!defOp && type == SourceKind::Unknown)
     // Check if the memory source is coming through a dummy argument.
     if (isDummyArgument(v)) {
-      type = SourceKind::Argument;
       ty = v.getType();
       if (fir::valueHasFirAttribute(v, fir::getTargetAttrName()))
         attributes.set(Attribute::Target);
-
       if (Source::isPointerReference(ty))
         attributes.set(Attribute::Pointer);
+      if (followBoxAddr && fir::isa_ref_type(ty))
```

Comment:
I'm not sure I understand your meaning.  Given the following:

```
subroutine test(p)
  real, pointer, intent(in) :: p
end subroutine test
```

flang-new generates:

```
func.func @_QPtest(%arg0: !fir.ref<!fir.box<!fir.ptr<f32>>> {fir.bindc_name = "p"}) {
  %0:2 = hlfir.declare %arg0 {fortran_attrs = #fir.var_attrs<intent_in, pointer>, uniq_name = "_QFtestEp"} : (!fir.ref<!fir.box<!fir.ptr<f32>>>) -> (!fir.ref<!fir.box<!fir.ptr<f32>>>, !fir.ref<!fir.box<!fir.ptr<f32>>>)
  return
}
```

If I drop the `intent(in)`, the only difference is that the `intent_in` attribute is lost.

Would you please provide an example of the case you have in mind?  Thanks.

Reply:
After further thought, maybe there is a cleaner way to write this check:

- Declare a new `followBoxAddrLoad` variable local to `AliasAnalysis::getSource`, and set it to true right before the `return` in the `fir::LoadOp` case.
- For the dummy arg check above, replace `followBoxAddr && fir::isa_ref_type(ty)` with `followBoxAddrLoad`.
- For the `fir::AddrOfOp` check, replace `followBoxAddr && mlir::isa<fir::BaseBoxType>(fir::unwrapRefType(ty))` with `followBoxAddrLoad`.

I tried that, and check-flang still succeeded.  What do you think?

Reply:
I am in the process of compiling your changes and will give them a better look today. What you describe though sounds really good. I will give it a shot.

Reply:
This is the reason why alias-analysis-host-assoc.fir is failing. 
This is the only place where we check for `SourceKind::Argument `


```
  if (src1->kind == SourceKind::Argument &&
      src2->kind == SourceKind::HostAssoc) {
   ...
    src2->attributes.set(Attribute::Target);
  }
```

Reply:
Are you saying you want `y(1)` in that case to be considered a `SourceKind::Direct` instead of a `SourceKind::Argument`?

Reply:
With your change, we are going to have to revisit how we detect a dummy argument. The SourceKind no longer captures it. Long story short, I think it is fine. We may want to also revisit the SourceKind nomenclature at some point but I think that's a different issue. When the search up the use-def chain reaches the same mlir::Value for a box and its data, we need some way of encoding that they are different and I guess the SourceKind is what we have used for that. 

The host association test still needs to know that it's dealing with a dummy and so instead of checking for SourceKind we can check for the mlir::Value. The check would become something like
```
if ((src1->isDummyArgument() && src2->kind == SourceKind::HostAssoc) || (src2->isDummyArgument() && src1->kind)
```

The `fir::isa_ref_type(ty)` restriction seems arbitrary. I understand that it helps with this test and limits updates of the tbaa  tests but ultimately I would think you would want to apply it unrestricted to rip all the benefits. I will leave it up to you and Tom to decide which way to go since you are going to be more impacted. And maybe @vzakhari can weigh in since he runs benchmarks on our side.

For context, @vzakhari is also going to work on a change related to dummy arguments so tbaa can be properly applied after inlining. There, too, we will need more information about a dummy and SourceKind alone will be insufficient. So, generally speaking things are moving away from the SourceKing::Argument being really useful.

I am sorry, if I made any of this tedious. I look at this code every 6 month maybe and need a refresh each time.

Reply:
> As far as I know, this patch only affects the case of an address extracted from a fortran pointer that is a dummy argument.

Sorry, I forgot that it's a bit more general than that: it also affects an address extracted from an allocatable that is a dummy argument.

Reply:
> What additional benefits do you mean?
Boxes are also passed by value. You would want to distinguish between them and their data as well. 
In this case `followBoxAddrLoad` would not be set.

Reply:
I was due for an example of box, passed by value:
```
subroutine test(p)
  real [,TARGET] :: p(:)
end subroutine test
```
```
func.func @_QPtest(%arg0: !fir.box<!fir.array<?xf32>> {fir.bindc_name = "p", fir.target})
```
I am not the expert here (that would be @jeanPerier ), but the rationale was to communicate in FIR that the callee does not modify the box and that therefore, to the caller, the box is a shallow constant. They do end up being passed by reference in codegen but show as values in FIR.

Reply:
> > That test doesn't deal with a fortran pointer, so this patch leaves it as SourceKind::Argument
> 
> `fir::isa_ref_type(ty)` does not mean fortran POINTER. `fir::isPointerType(ty)` does.

Right, when I wrote that comment, I somehow forgot that I generalized beyond pointers.  See my previous comment.
 
> > What additional benefits do you mean?
> 
> Boxes are also passed by value. You would want to distinguish between them and their data as well. In this case `followBoxAddrLoad` would not be set.

While the cases I'm trying to address so far specifically involve the `fir::LoadOp` case in `AliasAnalysis::getSource`, I'd appreciate any fortran example that does not.  My understanding of relevant use cases is probably too narrow.

Reply:
I wonder if it will be more clear to only keep the first four source kinds and move `Direct` into the attributes.  We are looking for the source entity here, so the source kind should represent this source entity.  Whatever happens during the use-def chain walk should not affect the kind of the source entity that we end up finding.  The `Direct`, in this case, would mean that the reference value passed to `getSource` is a result of one or more dereferences of the original source.  I just find it easier to think about it this way, though, you may know cases where it can complicate the implementation.

Reply:
> I was due for an example of box, passed by value:
> 
> ```
> subroutine test(p)
>   real [,TARGET] :: p(:)
> end subroutine test
> ```
> 
> ```
> func.func @_QPtest(%arg0: !fir.box<!fir.array<?xf32>> {fir.bindc_name = "p", fir.target})
> ```
> 
> I am not the expert here (that would be @jeanPerier ), but the rationale was to communicate in FIR that the callee does not modify the box and that therefore, to the caller, the box is a shallow constant. They do end up being passed by reference in codegen but show as values in FIR.

Thanks for the example.  That helps.  However, I'm struggling with the idea of performing a FIR alias analysis on something (the box in this example) that FIR represents as a value not an address.  If indeed it never makes sense to do that, then it seems it wouldn't matter whether we distinguish this box from its content as only the content is a candidate for alias analysis.

> I wonder if it will be more clear to only keep the first four source kinds and move `Direct` into the attributes. We are looking for the source entity here, so the source kind should represent this source entity. Whatever happens during the use-def chain walk should not affect the kind of the source entity that we end up finding. The `Direct`, in this case, would mean that the reference value passed to `getSource` is a result of one or more dereferences of the original source. I just find it easier to think about it this way, though, you may know cases where it can complicate the implementation.

Might make sense.  However, I've been assuming that, once you extract an address from a box, then you no longer have a global, dummy arg, local, or whatever the box is, so I thought it was ok to just call it a `Direct`.  Maybe that's wrong?  On that point, does anyone object to this patch's changes to `flang/test/Transforms/tbaa2.fir`?  There, this patch causes the address extracted from a dummy argument allocatable to be labeled `SourceKind::Direct` and not `SourceKind::Argument`.  Is that problematic?

Reply:
> I'm struggling with the idea of performing a FIR alias analysis on something (the box in this example) that FIR represents as a value not an address

Right, we could do some ground work before adding further enhancements to enforce that we start from a memory reference always. I wanted to check with @jeanPerier before we do that, in case he might need to make adjustments in lowering. 

Let's note that when we are at the LLVM level, the box is an address and it becomes a legitimate question to ask. I will have to dig some to see how we handle this case.

What I would be struggling with is that following the address of a POINTER takes you to a SourceKind::Direct but following the address of a TARGET takes you to a SourceKind::Argument.

> I wonder if it will be more clear to only keep the first four source kinds and move Direct into the attributes

Yes, then we would get consistent source kinds for target and pointers and we should get the same attribute, that I would hope we would have renamed to something other than "Direct" then. That would require some buy-in from @tblah as it would require changes in `runOnAliasInterface` though it should not cause any change in the results.

Reply:
Talking to Jean, reminded me that we are already handling the query on a box value as a query on the data. 
`bool followBoxAddr{mlir::isa<fir::BaseBoxType>(v.getType())};` 
We initialize `followBoxAddr` to true if we are starting from a box.

Though this is a moot point for this discussion since you are interested on a distinction between references.

I am working on a proposal to eliminate SourceKind::Direct while still retaining the information it was capturing. You still will be able to distinguish between data and box. I will write up an RFC tomorrow and will send out a WIP PR.